File: lab/snippet.py
"""
Module containing classes for creating merge requests
"""

# SPDX-FileCopyrightText: 2020 Jonah Brüchert <jbb@kaidan.im>
#
# SPDX-License-Identifier: GPL-2.0-or-later

import argparse
import sys

from typing import TextIO, Optional

from gitlab.v4.objects import Snippet

from lab.repositoryconnection import RepositoryConnection
from lab.utils import Utils, LogType


def parser(
    subparsers: argparse._SubParsersAction,  # pylint: disable=protected-access
) -> argparse.ArgumentParser:
    """
    Subparser for paste command
    :param subparsers: subparsers object from global parser
    :return: merge request creation subparser
    """
    snippet_parser: argparse.ArgumentParser = subparsers.add_parser(
        "snippet", help="Create a snippet from stdin or file", aliases=["paste"]
    )
    snippet_parser.add_argument("--title", help="Add a custom title", default="Empty title")
    snippet_parser.add_argument(
        "filename",
        metavar="str",
        type=str,
        nargs="?",
        help="File name to uplad",
    )
    return snippet_parser


def run(args: argparse.Namespace) -> None:
    """
    run snippet creation commands
    :param args: parsed arguments
    """
    snippets = Snippets()
    file: TextIO
    if args.filename:
        try:
            file = open(args.filename, "r")
        except FileNotFoundError:
            Utils.log(LogType.ERROR, "Failed to open file", args.filename)
            sys.exit(1)
    else:
        file = sys.stdin

    snippets.paste(file, title=args.title)


class Snippets(RepositoryConnection):
    """
    Class for creating snippets
    """

    def __init__(self) -> None:
        RepositoryConnection.__init__(self)

    def paste(self, file: TextIO, title: Optional[str]) -> None:
        """
        paste the contents of a TextIO object
        """
        snippet: Snippet = self._connection.snippets.create(
            {"title": title, "file_name": file.name, "content": file.read(), "visibility": "public"}
        )
        Utils.log(LogType.INFO, "Created snippet at", snippet.web_url)
        print("You can access it raw at", snippet.raw_url)

File: src/snippet.cpp
#include "snippet.h"

#include <map>
#include <iostream>
#include <fstream>

#include <pybind11/stl.h>
#include <pybind11/stl_bind.h>

#include "utils.h"

Snippet::Snippet() : RepositoryConnection()
{
}

void Snippet::run(const std::optional<std::string> &filename, const std::optional<std::string> &title)
{
    Snippet snippet;

    std::string content;
    if (filename) {
        auto file = std::ifstream(*filename);
        if (!file.is_open()) {
            Utils::log(LogType::Error, "Failed to open file" + *filename);
            std::exit(1);
        }
        for (std::string line; std::getline(file, line); ) {
            content += line;
        }
    } else {
        std::cin >> content;
    }

    snippet.paste(filename.value_or("stdin"), content, title.value_or("Empty title"));
}

void Snippet::paste(const std::string &file_name, const std::string &content, const std::string &title)
{
    std::map<std::string, std::string> options {
        {"title", title},
        {"file_name", file_name},
        {"content", content},
        {"visibility", "public"}
    };

    auto snippet = m_connection.attr("snippets").attr("create")(options);

    Utils::log(LogType::Info, "Created snippet at " + snippet.attr("web_url").cast<std::string>());
    std::cout << "You can access it raw at " << snippet.attr("raw_url").cast<std::string>();
}
